File: network_frame_cpp/src/transport/udp.cpp
#include "transport/udp.h"
#include <sstream>
#include <iomanip>

namespace transport {

UDPDatagram::UDPDatagram(uint16_t srcPort, uint16_t dstPort, const std::vector<uint8_t>& payload)
    : payload_(payload) {
    header_.srcPort = srcPort;
    header_.dstPort = dstPort;
    header_.length = static_cast<uint16_t>(UDP_HEADER_SIZE + payload.size());
    header_.checksum = 0;  // 校验和置0
}

std::vector<uint8_t> UDPDatagram::encode() const {
    std::vector<uint8_t> buf(UDP_HEADER_SIZE + payload_.size());
    
    // 写入UDP首部（大端/网络字节序）
    buf[0] = (header_.srcPort >> 8) & 0xFF;
    buf[1] = header_.srcPort & 0xFF;
    buf[2] = (header_.dstPort >> 8) & 0xFF;
    buf[3] = header_.dstPort & 0xFF;
    buf[4] = (header_.length >> 8) & 0xFF;
    buf[5] = header_.length & 0xFF;
    buf[6] = (header_.checksum >> 8) & 0xFF;
    buf[7] = header_.checksum & 0xFF;
    
    // 写入数据
    std::copy(payload_.begin(), payload_.end(), buf.begin() + UDP_HEADER_SIZE);
    
    return buf;
}

UDPDatagram UDPDatagram::decode(const std::vector<uint8_t>& data) {
    if (data.size() < UDP_HEADER_SIZE) {
        throw std::runtime_error("Data too short for UDP header");
    }
    
    UDPDatagram datagram;
    
    // 解析首部
    datagram.header_.srcPort = (static_cast<uint16_t>(data[0]) << 8) | data[1];
    datagram.header_.dstPort = (static_cast<uint16_t>(data[2]) << 8) | data[3];
    datagram.header_.length = (static_cast<uint16_t>(data[4]) << 8) | data[5];
    datagram.header_.checksum = (static_cast<uint16_t>(data[6]) << 8) | data[7];
    
    size_t payloadLen = datagram.header_.length - UDP_HEADER_SIZE;
    if (data.size() < UDP_HEADER_SIZE + payloadLen) {
        throw std::runtime_error("Invalid UDP length field");
    }
    
    datagram.payload_.assign(data.begin() + UDP_HEADER_SIZE, 
                              data.begin() + UDP_HEADER_SIZE + payloadLen);
    
    return datagram;
}

std::vector<uint8_t> UDPDatagram::getPayload() const {
    return payload_;
}

UDPHeader UDPDatagram::getHeader() const {
    return header_;
}

std::string UDPDatagram::toString() const {
    std::ostringstream oss;
    oss << "UDP Datagram:\n"
        << "  Source Port: " << header_.srcPort << "\n"
        << "  Dest Port: " << header_.dstPort << "\n"
        << "  Length: " << header_.length << "\n"
        << "  Checksum: 0x" << std::hex << std::setfill('0') << std::setw(4) << header_.checksum << "\n"
        << "  Payload Size: " << std::dec << payload_.size() << " bytes";
    return oss.str();
}

} // namespace transport

File: network_frame/pkg/transport/udp.go
package transport

import (
	"encoding/binary"
	"errors"
	"fmt"
)

const (
	// UDPHeaderSize UDP首部长度为8字节
	UDPHeaderSize = 8
)

// UDPHeader UDP首部结构
type UDPHeader struct {
	SrcPort  uint16 // 源端口
	DstPort  uint16 // 目的端口
	Length   uint16 // UDP长度（首部+数据）
	Checksum uint16 // 校验和（可选，置0）
}

// UDPDatagram UDP数据报
type UDPDatagram struct {
	Header  UDPHeader
	Payload []byte
}

// NewUDPDatagram 创建新的UDP数据报
func NewUDPDatagram(srcPort, dstPort uint16, payload []byte) *UDPDatagram {
	return &UDPDatagram{
		Header: UDPHeader{
			SrcPort:  srcPort,
			DstPort:  dstPort,
			Length:   uint16(UDPHeaderSize + len(payload)),
			Checksum: 0,
		},
		Payload: payload,
	}
}

// Encode 将UDP数据报编码为字节切片
func (u *UDPDatagram) Encode() []byte {
	buf := make([]byte, UDPHeaderSize+len(u.Payload))

	binary.BigEndian.PutUint16(buf[0:2], u.Header.SrcPort)
	binary.BigEndian.PutUint16(buf[2:4], u.Header.DstPort)
	binary.BigEndian.PutUint16(buf[4:6], u.Header.Length)
	binary.BigEndian.PutUint16(buf[6:8], u.Header.Checksum)

	copy(buf[UDPHeaderSize:], u.Payload)

	return buf
}

// DecodeUDP 从字节切片解码UDP数据报
func DecodeUDP(data []byte) (*UDPDatagram, error) {
	if len(data) < UDPHeaderSize {
		return nil, errors.New("data too short for UDP header")
	}

	header := UDPHeader{
		SrcPort:  binary.BigEndian.Uint16(data[0:2]),
		DstPort:  binary.BigEndian.Uint16(data[2:4]),
		Length:   binary.BigEndian.Uint16(data[4:6]),
		Checksum: binary.BigEndian.Uint16(data[6:8]),
	}

	payloadLen := int(header.Length) - UDPHeaderSize
	if payloadLen < 0 || len(data) < UDPHeaderSize+payloadLen {
		return nil, errors.New("invalid UDP length field")
	}

	payload := make([]byte, payloadLen)
	copy(payload, data[UDPHeaderSize:UDPHeaderSize+payloadLen])

	return &UDPDatagram{
		Header:  header,
		Payload: payload,
	}, nil
}

// String 返回UDP数据报的字符串表示
func (u *UDPDatagram) String() string {
	return fmt.Sprintf("UDP Datagram:\n"+
		"  Source Port: %d\n"+
		"  Dest Port: %d\n"+
		"  Length: %d\n"+
		"  Checksum: 0x%04X\n"+
		"  Payload Size: %d bytes",
		u.Header.SrcPort,
		u.Header.DstPort,
		u.Header.Length,
		u.Header.Checksum,
		len(u.Payload))
}

// GetPayload 获取UDP数据报的有效载荷
func (u *UDPDatagram) GetPayload() []byte {
	return u.Payload
}
